### PR TITLE
Add support for animate string value with %

### PR DESCRIPTION
**Pull request by piaskowyk**
## Description

Add support for animate string value with %
Fixes: https://github.com/software-mansion/react-native-reanimated/issues/1613

## Example code
<details>
<summary>Example code</summary>

```js
import Animated, {
  useSharedValue,
  withTiming,
  useAnimatedStyle,
  Easing,
} from 'react-native-reanimated';
import { View, Button } from 'react-native';
import React from 'react';

export default function AnimatedStyleUpdateExample(props) {
  const randomWidth = useSharedValue('10%');
  
  const config = {
    duration: 500,
    easing: Easing.bezier(0.5, 0.01, 0, 1),
  };

  const style = useAnimatedStyle(() => {
    return {
      width: withTiming(randomWidth.value, config),
    };
  });

  return (
    <View>
      <Animated.View
        style={[
          { width: '100%', height: 80, backgroundColor: 'black' },
          style,
        ]}
      />
      <Button
        title="toggle"
        onPress={() => {
          randomWidth.value = Math.random() * 100 + '%' ;
        }}
      />
    </View>
  );
}

```

</details>

### CODE REVIEW

File: src/reanimated2/animations.js [hunk 1]
```diff
@@ -48,6 +48,34 @@ export function transformAnimation(animation) {
...
+  }
+  animation.toValue = animation.toValue + unit;
+  animation.current = animation.current + unit;
+  animation.startValue = animation.startValue + unit;
+}
+
+export function getUnit(value) {
+  'worklet';
+  for (const unit of ['%', 'pt']) {
+    if (value.endsWith(unit)) return unit;
```
Comment (jakub-gonet) [inline src/reanimated2/animations.js:74]: Missing braces.

File: src/reanimated2/animations.js [hunk 2]
```diff
@@ -82,6 +110,21 @@ export function decorateAnimation(animation) {
     return res;
   };
 
+  const withUnitOnStart = (animation, value, timestamp, previousAnimation) => {
+    const val = parseFloat(value);
+    removeAnimationUnit(animation);
+    if (previousAnimation !== animation) removeAnimationUnit(previousAnimation);
```
Comment (jakub-gonet) [inline src/reanimated2/animations.js:116]: Missing braces.

**Review by jakub-gonet — COMMENTED** [review]
Could you add some examples, update TS types if necessary, and update docs?

File: src/reanimated2/animations.js [hunk 3]
```diff
@@ -48,6 +48,36 @@ export function transformAnimation(animation) {
...
+    return;
+  }
+  animation.toValue = animation.toValue + unit;
+  animation.current = animation.current + unit;
+  animation.startValue = animation.startValue + unit;
+}
+
+export function getUnit(value) {
+  'worklet';
+  for (const unit of ['%', 'pt']) {
```
Comment (karol-bisztyga) [inline src/reanimated2/animations.js:73]: maybe it would be good to define this array somewhere as, for instance, `ALLOWED_UNITS`?

**Review by karol-bisztyga — COMMENTED** [review]
Looks alright.
Could you please add an example where the `'pt'` unit occurs?
Maybe @Szymon20000 could do a code review here?

File: src/reanimated2/animations.js (hunk 3 above)
Comment (piaskowyk) [reply src/reanimated2/animations.js:73]: `pt` works only for web. RN supports only `%` as unit, but RN for web supports [`%`, `pt`, `px`, `em`, `rem`, `vh`, `vw`] So what do you think, add all of this as ALLOWED_UNITS or limit support only for `%` for all platforms?

Comment (karol-bisztyga) [reply src/reanimated2/animations.js:73]: In such a case I would go for conditional contents of allowed units.
Something like this maybe:
```
const ALLOWED_UNITS = Platform.OS === 'web' ? ['%', 'pt' ...] : ['%'];
```

File: react-native-reanimated.d.ts [hunk 4]
```diff
@@ -403,12 +403,12 @@ declare module 'react-native-reanimated' {
       easing?: EasingFunction;
     }
     export function withTiming(
-      toValue: number | Exclude<ColorValue, OpaqueColorValue>,  // string as a color value like `"rgba(20,20,20,0)"`
+      toValue: number | string | Exclude<ColorValue, OpaqueColorValue>,  // string as a color value like `"rgba(20,20,20,0)"`
```
Comment (jakub-gonet) [inline react-native-reanimated.d.ts:406]: `Exclude<ColorValue, OpaqueColorValue>` already contains `string`

File: src/reanimated2/animations.js [hunk 5]
```diff
@@ -21,7 +21,7 @@ export function transform(value, handler) {
   if (typeof value === 'string') {
     // toInt
     // TODO handle color
-    const match = value.match(/([A-Za-z]*)([-\d.]*)([A-Za-z]*)/);
+    const match = value.match(/([A-Za-z]*)([-\d.]*)([A-Za-z%]*)/);
```
Comment (jakub-gonet) [inline src/reanimated2/animations.js:24]: This regex is wrong, `([-\d.]*)` allows parsing `-1-...1` as a number.
I think `(-?\d*\.?\d*)` is slightly better, it matches the optional minus sign, and optional dot between numbers (see attached screenshot).
image

**Comment by jakub-gonet** [comment]
Please don't force push into branches that have a pull request and reviews. It removes old commits, so you can't check history at the time of writing comments.

**Review by karol-bisztyga — APPROVED** [review]

**Review by jakub-gonet — APPROVED** [review]

**Pull request merged**